### PR TITLE
SAP propagate scheduler hints for volumes

### PR DESCRIPTION
**Pull request by hemna**
This patch ensures that when a new volume is create and if it has scheduler hints for affinity/anti-affinity for other volumes, that the volumes get affinity/anti-affinity for the newly created volumes.  This ensures affinity and anti-affinity are both ways for volumes.

### CODE REVIEW

File: cinder/volume/manager.py [hunk 1]
```diff
@@ -680,6 +680,43 @@ def _set_resource_host(self, resource):
             resource.host = volume_utils.append_host(self.host, pool)
             resource.save()
 
+    def _propogate_volume_scheduler_hints(self, context, volume):
+        """Ensure metadata hints are propagated to referenced volumes."""
+        hints = volume_utils.get_scheduler_hints_from_volume(volume)
+        if hints:
```
Comment (joker-at-work) [inline cinder/volume/manager.py:686]:
```python
if not hints:
    return
```
would look much better here - at least to me ;)

File: cinder/volume/manager.py [hunk 2]
```diff
@@ -680,6 +680,43 @@ def _set_resource_host(self, resource):
...
+            LOG.debug("Found hints for %(volume)s - %(hints)s",
+                      {'volume': volume.id, 'hints': hints})
+            for hint_key in hints.keys():
```
Comment (joker-at-work) [inline cinder/volume/manager.py:689]: Why do you always use `.keys()` when it's the default behavior of a `dict`? Is there any specific reason to it? It's a little odd for me, so I'd like to understand.

File: cinder/volume/manager.py [hunk 3]
```diff
@@ -680,6 +680,43 @@ def _set_resource_host(self, resource):
...
+                        LOG.debug("Set hints in metadata %s", subhints)
+                        md = vut.set_scheduler_hints_to_volume_metadata(
+                            subhints, meta_vol.metadata
+                        )
+                        LOG.debug("metadata  %s", md)
+                        meta_vol.metadata = md
+                        meta_vol.save()
+                    except Exception as ex:
+                        LOG.error("Failed to set scheduler hints")
+                        LOG.error(ex)
```
Comment (joker-at-work) [inline cinder/volume/manager.py:718]: This could be a single `LOG.exception("Failed to set scheduler hints")`, I guess. Any reason for not printing the stacktrace and just logging the exception message i.e. the stringified exception - in another line?

File: cinder/volume/manager.py [hunk 4]
```diff
@@ -680,6 +680,43 @@ def _set_resource_host(self, resource):
...
+                            else:
+                                subhints[hint_key] = [volume.id]
+                        else:
+                            subhints = {hint_key: [volume.id]}
+
+                        LOG.debug("Set hints in metadata %s", subhints)
+                        md = vut.set_scheduler_hints_to_volume_metadata(
+                            subhints, meta_vol.metadata
+                        )
+                        LOG.debug("metadata  %s", md)
```
Comment (joker-at-work) [inline cinder/volume/manager.py:713]: I'm not sure about all the debug messages above. They look a little like printf-debugging during development. Could we maybe do one, proper message with the information we would want to have?

File: cinder/volume/manager.py [hunk 5]
```diff
@@ -680,6 +680,43 @@ def _set_resource_host(self, resource):
...
+            LOG.debug("Found hints for %(volume)s - %(hints)s",
+                      {'volume': volume.id, 'hints': hints})
+            for hint_key in hints.keys():
+                for vol_id in hints[hint_key]:
+                    try:
+                        meta_vol = objects.Volume.get_by_id(context, vol_id)
+                        LOG.debug("Looking for hints in %s", meta_vol.id)
+                        # Because pep8 error otherwise with lines 696 and 710
+                        vut = volume_utils
+                        subhints = vut.get_scheduler_hints_from_volume(
```
Comment (joker-at-work) [inline cinder/volume/manager.py:696]: So ... `subhints` are the hints already existing on the `meta_vol`. Imho better called `meta_vol_hints` or `meta_hints`.

File: cinder/volume/manager.py [hunk 6]
```diff
@@ -680,6 +680,43 @@ def _set_resource_host(self, resource):
...
+                        )
+                        LOG.debug("hints? %s", subhints)
+                        if subhints:
+                            if (hint_key in subhints and
+                                    volume.id not in subhints[hint_key]):
+                                subhints[hint_key].append(volume.id)
+                            else:
+                                subhints[hint_key] = [volume.id]
+                        else:
+                            subhints = {hint_key: [volume.id]}
```
Comment (joker-at-work) [inline cinder/volume/manager.py:707]: `subhints` is always a `dict` - may be an empty one, though - so we can shorten this to
```python
subhints.setdefault(hint_key, [])
if volume.id not in subhints[hint_key]:
    subhints[hint_key].append(volume.id)
```

File: cinder/volume/manager.py (hunk 1 above)
Comment (hemna) [reply cinder/volume/manager.py:686]: done

File: cinder/volume/manager.py (hunk 2 above)
Comment (hemna) [reply cinder/volume/manager.py:689]: probably a remnant of old style

File: cinder/volume/manager.py (hunk 4 above)
Comment (hemna) [reply cinder/volume/manager.py:713]: I've removed them

File: cinder/volume/manager.py (hunk 5 above)
Comment (hemna) [reply cinder/volume/manager.py:696]: done

File: cinder/volume/manager.py (hunk 6 above)
Comment (hemna) [reply cinder/volume/manager.py:707]: done

File: cinder/volume/manager.py [hunk 7]
```diff
@@ -680,6 +680,36 @@ def _set_resource_host(self, resource):
...
+        if not hints:
+            return
+
+        LOG.debug("Found hints for %(volume)s - %(hints)s",
+                  {'volume': volume.id, 'hints': hints})
+        for hint_key in hints:
+            for vol_id in hints[hint_key]:
+                try:
+                    meta_vol = objects.Volume.get_by_id(context, vol_id)
+                    # Because pep8 error otherwise with lines 696 and 710
```
Comment (joker-at-work) [inline cinder/volume/manager.py:695]: The lines don't match anymore. Maybe it would be best to not give explicit line-numbers and just refer to "further down"?

File: cinder/volume/manager.py [hunk 8]
```diff
@@ -680,6 +680,36 @@ def _set_resource_host(self, resource):
...
+                        meta_vol_hints[hint_key].append(volume.id)
+
+                    md = vut.set_scheduler_hints_to_volume_metadata(
+                        meta_vol_hints, meta_vol.metadata
+                    )
+                    meta_vol.metadata = md
+                    meta_vol.save()
+                except Exception:
+                    LOG.error("Failed to set scheduler hints",
+                              resource=meta_vol)
```
Comment (joker-at-work) [inline cinder/volume/manager.py:711]: I think it would be better to use `LOG.exception()` here, so we can debug the problem properly.

Comment (hemna) [reply cinder/volume/manager.py:711]: done

File: cinder/volume/manager.py (hunk 7 above)
Comment (hemna) [reply cinder/volume/manager.py:695]: done

**Review by joker-at-work — APPROVED** [review]
